### PR TITLE
Account for relative_url_root in FailureApp's recall method

### LINKED ISSUE #3738: `failure_app::recall` doesn't take into account relative_url_root
When a login failure occurs, Warden sends back a full path in `attempted_path`, but devise's `failure_app` assumes the path does not contain a relative URL root and assigns it to `PATH_INFO`:

```ruby
  def recall
      env["PATH_INFO"]  = attempted_path
      flash.now[:alert] = i18n_message(:invalid) if is_flashing_format?
      self.response = recall_app(warden_options[:recall]).call(env)
    end
```

For example, suppose:

1.  `relative_url_root` = `/base_app`  
2. `attempted_path` = `/base_app/resource`.
3. In this case, `PATH_INFO` is assigned `/base_app/resource`.
4. However, `SessionController::store_redirect_path` will call `request.fullpath`, which then makes the path `/base_app/base_app/resource`. Note the redundant `base_app`.

`PATH_INFO` should only contain the part of the path that excludes the relative URL root, since `SCRIPT_NAME` includes the relative URL part (see http://api.rubyonrails.org/classes/ActionDispatch/Routing/UrlFor.html) for more details.

### PR DESCRIPTION
**Pull request by stanhu**
Closes #3738

### CODE REVIEW

**Comment by MichielDeMey** [comment]
+1

**Comment by stanhu** [comment]
@josevalim Can you please review this? Thank you.

**Comment by chrismacias** [comment]
it seems that this fixes the issue

**Comment by stanhu** [comment]
Friendly ping: @carlosantoniodasilva @josevalim, can one of you review?

File: test/failure_app_test.rb [hunk 1]
```diff
@@ -294,5 +294,22 @@ def call_failure(env_params={})
       assert @response.third.body.include?('<h2>Log in</h2>')
       assert @response.third.body.include?('Your account is not activated yet.')
     end
+
+    if Rails.application.config.respond_to?(:relative_url_root)
```
Comment (josevalim) [inline test/failure_app_test.rb]: Is this required to support different rails versions?

File: lib/devise/failure_app.rb [hunk 2]
```diff
@@ -51,7 +51,18 @@ def http_auth
     end
 
     def recall
-      env["PATH_INFO"]  = attempted_path
+      config = Rails.application.config
+
+      if config.try(:relative_url_root)
+        base_path = Pathname.new(config.relative_url_root)
```
Comment (josevalim) [inline lib/devise/failure_app.rb]: Can we not use `Pathname` and call the methods in `File` or `Path` directly? I always find `Pathname` very error prone, specially if you forget to call `to_s`.

**Comment by josevalim** [comment]
Sorry for the delay! I have added some comments. :) Thank you!

File: lib/devise/failure_app.rb (hunk 2 above)
Comment (stanhu) [inline lib/devise/failure_app.rb]: The method makes use of Pathname's `relative_path_from` method. Is there a better substitute? As far as I can tell, there isn't anything else.

File: test/failure_app_test.rb (hunk 1 above)
Comment (stanhu) [inline test/failure_app_test.rb]: I was under the impression `relative_url_root` could be left blank in the application config.

File: lib/devise/failure_app.rb (hunk 2 above)
Comment (josevalim) [inline lib/devise/failure_app.rb]: You are right, thank you.

File: test/failure_app_test.rb (hunk 1 above)
Comment (josevalim) [inline test/failure_app_test.rb]: Yes, but does it matter for Devise own test suite?

Comment (stanhu) [inline test/failure_app_test.rb]: There are a number of test cases that use this pattern. Should I remove them all?

Comment (josevalim) [inline test/failure_app_test.rb]: If there are other tests relying on this pattern you have answered my question, it is for Rails compat. We can probably remove it in another PR if the build passes after doing so. :)

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by stanhu** [comment]
Thanks, @josevalim!

**Pull request merged**